pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,434 @@
+	dbClient, err := client.NewDatabase(params.Database)
+	if err != nil {
+		return nil, fmt.Errorf("creating database client: %w", err)
+	}
+	dbResp, err := dbClient.Read(ctx, nil)
+	errCode := errStatusCode(err)
+	switch {
+	case errCode == -1 && dbResp.RawResponse != nil && dbResp.RawResponse.StatusCode == http.StatusOK:
+		return dbClient, nil
+	case errCode == http.StatusNotFound:
+		dbResp, err = client.CreateDatabase(ctx, azcosmos.DatabaseProperties{ID: params.Database}, nil)
+		if err != nil || dbResp.RawResponse.StatusCode != http.StatusCreated {
+			return nil, fmt.Errorf("reading database(%d): %w", dbResp.RawResponse.StatusCode, err)
+		}
+	default:
+		return nil, fmt.Errorf("reading database: %w", err)
+	}
+	return dbClient, nil
+}
+
+func getOrCreateContainer(ctx context.Context, dbClient *azcosmos.DatabaseClient, params *kvparams.CosmosDB) (*azcosmos.ContainerClient, error) {
+	containerClient, err := dbClient.NewContainer(params.Container)
+	if err != nil {
+		return nil, fmt.Errorf("creating database client: %w", err)
+	}
+	cResp, err := containerClient.Read(ctx, nil)
+	errCode := errStatusCode(err)
+	switch {
+	case errCode == -1 && cResp.RawResponse != nil && cResp.RawResponse.StatusCode == http.StatusOK:
+		return containerClient, nil
+	case errCode == http.StatusNotFound:
+		cResp, err = dbClient.CreateContainer(ctx,
+			azcosmos.ContainerProperties{
+				ID: params.Container,
+				PartitionKeyDefinition: azcosmos.PartitionKeyDefinition{
+					Paths: []string{"/partitionKey"},
+				},
+				// Excluding the value field from indexing since it is not used in queries and saves RUs for writes.
+				// partitionKey is automatically not indexed. The rest of the fields are indexed by default, including id
+				// which is unnecessary, but cannot be excluded.
+				IndexingPolicy: &azcosmos.IndexingPolicy{
+					Automatic:     false,
+					IndexingMode:  azcosmos.IndexingModeConsistent,
+					IncludedPaths: []azcosmos.IncludedPath{{Path: "/*"}},
+					ExcludedPaths: []azcosmos.ExcludedPath{{Path: "/value/?"}},
+				},
+			}, nil)
+
+		if err != nil || cResp.RawResponse.StatusCode != http.StatusCreated {
+			return nil, fmt.Errorf("creating container: %w", err)
+		}
+	default:
+		return nil, fmt.Errorf("reading database(%d): %w", cResp.RawResponse.StatusCode, err)
+	}
+
+	return containerClient, nil
+}
+
+// encoding is the encoding used to encode the partition keys, ids and values.
+// Must be an encoding that keeps the strings in-order.
+var encoding = base32.HexEncoding // Encoding that keeps the strings in-order.
+
+// hashID returns a hash of the key that is used as the document id.
+func (s *Store) hashID(key []byte) string {
+	return encoding.EncodeToString(ident.NewAddressWriter().MarshalBytes(key).Identity())
+}
+
+type Document struct {
+	PartitionKey string `json:"partitionKey"`
+	HashID       string `json:"id"`
+	SortID       string `json:"sortID"`
```

Comment:
To align the names and make it clear I would go with `ID` for the internal unique identifier.
And key for our key/value.
Maybe add code doc to explain each one.
```suggestion
	ID       string `json:"id"`
	Key       string `json:"key"`
```

Reply:
Done.